src/Connection/Internal/Http2ConnectionProcessor.php

```diff
@@ -905,29 +918,34 @@ public function request(Request $request, CancellationToken $cancellationToken,
                 throw $exception;
             }
 
+            $originalCancellation = $cancellationToken;
+            if ($request->getTransferTimeout() > 0) {
+                $cancellationToken = new CombinedCancellationToken(
+                    $cancellationToken,
+                    new TimeoutCancellationToken($request->getTransferTimeout())
+                );
+            }
+
             $streamId = $this->streamId += 2; // Client streams should be odd-numbered, starting at 1.
 
             $this->streams[$streamId] = $http2stream = new Http2Stream(
                 $streamId,
                 $request,
                 $stream,
                 $cancellationToken,
+                $originalCancellation,
                 self::DEFAULT_WINDOW_SIZE,
                 $this->initialWindowSize
             );
 
-            if ($request->getTransferTimeout() > 0) {
-                // Cancellation token combined with timeout token should not be stored in $stream->cancellationToken,
-                // otherwise the timeout applies to the body transfer and pushes.
-                $cancellationToken = new CombinedCancellationToken(
-                    $cancellationToken,
-                    new TimeoutCancellationToken($request->getTransferTimeout())
-                );
-            }
```

Comment:
Seems like this commend has been introduced by you, @trowski: 93c2e6e5baab68558deb8c0c3e7b8131726ba1db

Reply:
I think my assumption at that time was that the transfer timeout only applied to the headers and that there was a separate timeout for transferring the response body.